src/allocation_engine/interval_tree.rs

```diff
@@ -185,8 +185,173 @@ struct InnerNode {
     right: Option<Box<InnerNode>>,
     /// Cached height of the node.
     height: u64,
-    /// Cached maximum valued covered by this node.
-    max_key: u64,
+}
+
+impl InnerNode {
+    /// Creates a new InnerNode object.
+    #[allow(dead_code)]
+    fn new(key: Range, node_state: NodeState) -> Self {
+        InnerNode {
+            key,
+            node_state,
+            left: None,
+            right: None,
+            height: 1,
+        }
+    }
+
+    /// Returns a readonly reference to the node associated with the `key` or
+    /// None if the searched key does not exist in the tree.
+    #[allow(dead_code)]
+    fn search(&self, key: &Range) -> Option<&InnerNode> {
+        match self.key.cmp(key) {
+            Ordering::Equal => Some(self),
+            Ordering::Less => self.right.as_ref().and_then(|node| node.search(key)),
+            Ordering::Greater => self.left.as_ref().and_then(|node| node.search(key)),
+        }
+    }
+
+    /// Returns a readonly reference to the node associated with the `key` or
+    /// None if there is no Node representing an interval that covers the
+    /// searched key. For a key [a, b], this method will return a node with
+    /// a key [c, d] such that c <= a and b <= d.
+    #[allow(dead_code)]
+    fn search_superset(&self, key: &Range) -> Option<&InnerNode> {
+        if self.key.contains(key) {
+            Some(self)
+        } else if key.end < self.key.start {
+            self.left
+                .as_ref()
+                .and_then(|node| node.search_superset(key))
+        } else {
+            self.right
+                .as_ref()
+                .and_then(|node| node.search_superset(key))
+        }
+    }
+
+    /// Rotates the tree such that height difference between subtrees
+    /// is not greater than abs(1).
+    #[allow(dead_code)]
+    fn rotate(self: Box<Self>) -> Box<Self> {
+        let l = height(&self.left);
+        let r = height(&self.right);
+
+        match (l as i64) - (r as i64) {
+            1 | 0 | -1 => self,
+            2 => self.rotate_left_successor(),
+            -2 => self.rotate_right_successor(),
+            _ => unreachable!(),
+        }
+    }
+
+    /// Performs a single left rotation on this node.
+    #[allow(dead_code)]
+    fn rotate_left(mut self: Box<Self>) -> Box<Self> {
+        let mut new_root = self.right.take().expect("Node is broken");
```

Comment:
Why is this unwrap safe? Can you add a comment? We shouldn't have unwraps in the code unless they can be caused only by programming errors.

Reply:
Reading the comments, I think it's actually safer (in the sense that it's less prone to mistakes in the future) if we just return an `Option` here, and do the unwrap in the function where we validated that the child exists. The risk is that someone makes this function public, and then it can be called without doing the previously mentioned validation.

Reply:
Okay. Will update it